support for S3 response header overrides

zvikagart opened the pull request:
I added support for `response-content-disposition` request query parameter in object GETs.
See details: https://forums.aws.amazon.com/ann.jspa?annID=884

I also added a simple Dockerfile to support a docker development environment based on the official [maven docker image](https://hub.docker.com/_/maven/). You can then run `docker build -t s3proxy-build -f Dockerfile.build .` to build and test s3proxy. This encapsulates java and maven dependencies quite nicely.

andrewgaul commented:
@zvikagart Can you add or modify a test for the `response-content-disposition` parameter?  Also please remove the Dockerfile from this pull request and can you review the more configurable one in #123?

zvikagart commented:
I tried to dig a little bit to find a way to test this query parameter, but it looks like jclouds doesn't support this yet (https://issues.apache.org/jira/browse/JCLOUDS-1016). If you have any pointers I'll gladly implement this.

As for the Dockerfile I added, I intentionally named it differently (`Dockefile.build`) since it serves a totally different role than the Dockerfile in #123. The one I added is just for building the project, not to run the service.

andrewgaul commented:
Could you use the AWS SDK instead of jclouds?  I believe you just need to add your header to `SignableRequest`:

https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/SignableRequest.html

You can find other tests using the AWS SDK in `S3AwsSdkTest`.

andrewgaul commented:
@zvikagart I added a Dockerfile so please add the test and rebase this PR so I can merge it!

zvikagart commented:
@andrewgaul, I generalized the code to support the full set of response header override options ([the list](http://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectGET.html#RESTObjectGET-requests)), added a test, squashed and rebased.

andrewgaul commented on src/main/java/org/gaul/s3proxy/S3ProxyHandler.java:
```diff
@@ -56,6 +56,7 @@
 import javax.xml.stream.XMLStreamException;
 import javax.xml.stream.XMLStreamWriter;
 
+import com.amazonaws.services.s3.model.ResponseHeaderOverrides;
```
Let's not add a run-time dependency on the SDK just to get these constants.

andrewgaul commented on Dockerfile.build:
```diff
@@ -0,0 +1,3 @@
+FROM maven:3-jdk-7-onbuild
+
+CMD ["mvn test"]
```
Why do we need this `Dockerfile.build` now that S3Proxy provides a `Dockerfile`?

zvikagart replied:
As I explained before this is for a docker based build and development environment.
If you don't think it's useful I'll remove it.

andrewgaul replied:
Let's remove this for now. This would be useful to get the s3-tests to run locally so I opened #146.

andrewgaul commented on src/main/java/org/gaul/s3proxy/S3ProxyHandler.java:
```diff
@@ -2365,21 +2372,54 @@ private void handleUploadPart(HttpServletRequest request,
+            response.addHeader(headerName, override);
+        } else {
+            response.addHeader(headerName, value);
+        }
```
Simplify with  `Objects.firstNonNull`.

andrewgaul commented:
Overall looks good, please address comments so I can merge this.

andrewgaul commented on src/main/java/org/gaul/s3proxy/S3ProxyHandler.java:
```diff
@@ -2388,9 +2428,15 @@ private static void addMetadataToResponse(HttpServletResponse response,
+            Date expires = contentMetadata.getExpires();
+            if (expires != null) {
+                response.addDateHeader(HttpHeaders.EXPIRES, expires.getTime());
+            }
```
Similar `Objects.firstNonNull`.

zvikagart replied:
This one is not straightforward. To avoid unnecessary evaluations you'll have to use the `first.or(supplier)` pattern, plus the fact that you want `expires.getTime()` not just `expires`. I don't see this becoming more readable than it currently is.

andrewgaul replied:
Agree, sorry I misread this.

zvikagart commented on src/main/java/org/gaul/s3proxy/S3ProxyHandler.java:
```diff
@@ -2365,21 +2371,48 @@ private void handleUploadPart(HttpServletRequest request,
+            String overrideHeaderName,
+            String value) {
+        String override = request.getParameter(overrideHeaderName);
+        response.addHeader(headerName, override != null ? override : value);
```
Note that while trying to use `Objects.firstNonNull` I got lots of `NullPointerException` in tests. Probably due to not having a fallback value when calling `Objects.firstNonNull`, probably because some object metadata values were not set.
Instead I used the ternary operator.

andrewgaul commented:
I made minor stylistic changes, rebased, and pushed to master as fcbc14780b3c45359553848727ac641b499b3e6b.  Thank you for your contribution @zvikagart!
